`PluginAnalyzer` tries to analyze non-Skyrim plugin files

# Bug Report

## Summary
Open the App, login and select Skyrim as a game to manage.

The manage process (getting the game ready to mod/launch) takes a very long time (minutes).

## Steps to reproduce
1. Install the app from the executable.
2. Launch the app.
3. Login.
4. Manage Skyrim (Steam?)


## What is the expected behaviour?
The process to get the game into a state ready to mod needs to be faster, it seems to be taking around 90 seconds.


## Other information
```
00:03:44.161 [ERROR] Error analyzing C:/Program Files (x86)/Steam/steamapps/common/Skyrim Special Edition/Tools/zEdit_v0.6.6.1_-_Portable_x64/Oblivion.Hardcoded.dat with PluginAnalyzer|System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   at Mutagen.Bethesda.Plugins.Binary.Translations.RecordSpanExtensions.EnumerateSubrecords(ReadOnlyMemorySlice`1 span, GameConstants meta, Int32 offset)+MoveNext()
   at Mutagen.Bethesda.HeaderExt.Masters(ModHeaderFrame modHeader)+MoveNext()
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Mutagen.Bethesda.Plugins.Masters.MasterReferenceCollection.SetTo(IEnumerable`1 masters)
   at Mutagen.Bethesda.Plugins.Masters.MasterReferenceCollection.FromStream[TStream](TStream stream)
   at Mutagen.Bethesda.Plugins.Masters.MasterReferenceCollection.FromStream(Stream stream, ModKey modKey, GameRelease release, Boolean disposeStream)
   at Mutagen.Bethesda.Plugins.Binary.Streams.MutagenBinaryReadStream..ctor(Stream stream, ModKey modKey, GameRelease release, Int32 bufferSize, Boolean dispose, Int64 offsetReference)
   at Mutagen.Bethesda.Skyrim.SkyrimMod.CreateFromBinaryOverlay(Stream stream, SkyrimRelease release, ModKey modKey)
   at NexusMods.Games.BethesdaGameStudios.PluginAnalyzer.AnalyzeAsync(FileAnalyzerInfo info, CancellationToken ct)+MoveNext() in /_/src/Games/NexusMods.Games.BethesdaGameStudios/PluginAnalyzer.cs:line 21
   at NexusMods.DataModel.ArchiveAnalyzer.AnalyzeFileInnerAsync(IStreamFactory sFn, CancellationToken token, Int32 level, Hash parent, Nullable`1 parentArchivePath, RelativePath parentPath, String fileName) in /_/src/NexusMods.DataModel/ArchiveAnalyzer.cs:line 226
   at NexusMods.DataModel.ArchiveAnalyzer.AnalyzeFileInnerAsync(IStreamFactory sFn, CancellationToken token, Int32 level, Hash parent, Nullable`1 parentArchivePath, RelativePath parentPath, String fileName) in /_/src/NexusMods.DataModel/ArchiveAnalyzer.cs:line 226
00:05:14.591 [INFO] Loadout 19DD2B5D6C60C64F9534C383D2F89FB2 altered: Add game files
00:05:14.612 [INFO] Ready to show NexusMods.App.UI.LeftMenu.Items.IconView with autowired NexusMods.App.UI.LeftMenu.Items.IconViewModel.
```

--- a/src/Games/NexusMods.Games.BethesdaGameStudios/PluginAnalyzer.cs
+++ b/src/Games/NexusMods.Games.BethesdaGameStudios/PluginAnalyzer.cs
@@ -16,7 +16,7 @@ using Noggog;
 
 namespace NexusMods.Games.BethesdaGameStudios;
 
-[PublicAPI]
+[UsedImplicitly]
 public class PluginAnalyzer : IFileAnalyzer
 {
     public FileAnalyzerId Id { get; } = FileAnalyzerId.New("9c673a4f-064f-4b1e-83e3-4bf0454575cd", 1);

--- a/src/Games/NexusMods.Games.BethesdaGameStudios/PluginAnalyzer.cs
+++ b/src/Games/NexusMods.Games.BethesdaGameStudios/PluginAnalyzer.cs
@@ -1,30 +1,109 @@
 using System.Runtime.CompilerServices;
+using JetBrains.Annotations;
+using Microsoft.Extensions.Logging;
+using Mutagen.Bethesda;
 using Mutagen.Bethesda.Plugins;
+using Mutagen.Bethesda.Plugins.Binary.Streams;
+using Mutagen.Bethesda.Plugins.Meta;
+using Mutagen.Bethesda.Plugins.Records;
 using Mutagen.Bethesda.Skyrim;
 using NexusMods.DataModel.Abstractions;
 using NexusMods.DataModel.Abstractions.Ids;
 using NexusMods.FileExtractor.FileSignatures;
+using NexusMods.Paths;
 using NexusMods.Paths.Extensions;
+using Noggog;
 
 namespace NexusMods.Games.BethesdaGameStudios;
 
+[PublicAPI]
 public class PluginAnalyzer : IFileAnalyzer
 {
     public FileAnalyzerId Id { get; } = FileAnalyzerId.New("9c673a4f-064f-4b1e-83e3-4bf0454575cd", 1);
 
     public IEnumerable<FileType> FileTypes => new[] { FileType.TES4 };
 
+    private static readonly Extension[] ValidExtensions = {
+        new(".esp"),
+        new(".esm"),
+        new(".esl"),
+    };
+
+    private readonly ILogger<PluginAnalyzer> _logger;
+
+    public PluginAnalyzer(ILogger<PluginAnalyzer> logger)
+    {
+        _logger = logger;
+    }
+
 #pragma warning disable CS1998
     public async IAsyncEnumerable<IFileAnalysisData> AnalyzeAsync(FileAnalyzerInfo info, [EnumeratorCancellation] CancellationToken ct = default)
 #pragma warning restore CS1998
     {
-        var overlay = SkyrimMod.CreateFromBinaryOverlay(info.Stream, SkyrimRelease.SkyrimSE, new ModKey("Dummy", ModType.Plugin));
+        var extension = info.FileName.ToRelativePath().Extension;
+        if (ValidExtensions[0] != extension && ValidExtensions[1] != extension && ValidExtensions[2] != extension) yield break;
 
-        var masters = overlay.ModHeader.MasterReferences.Select(x => x.Master.FileName.ToString());
-        yield return new PluginAnalysisData
+        // NOTE(erri120): The GameConstant specifies the header length.
+        // - Oblivion: 20 bytes
+        // - Skyrim (every release): 24 bytes
+        // - Fallout 4: 24 bytes
+        // We're only looking for the "Light Master" flag and all Master References.
+        // The Flags are at 8+4 bytes, so we can use the smallest header length for those.
+        // However, the Master References are just Subrecords that appear directly after
+        // the header.
+        // The current solution just tries different GameConstants, which isn't ideal and
+        // should be replaced with an identification step that finds the correct GameConstant.
+
+        var startPos = info.Stream.Position;
+        var fileAnalysisData = Analyze(GameConstants.SkyrimLE, info);
+        if (fileAnalysisData is null)
         {
-            IsLightMaster = overlay.ModHeader.Flags.HasFlag(SkyrimModHeader.HeaderFlag.LightMaster),
-            Masters = masters.Select(m => m.ToRelativePath()).ToArray()
-        };
+            info.Stream.Position = startPos;
+            fileAnalysisData = Analyze(GameConstants.Oblivion, info);
+            if (fileAnalysisData is null) yield break;
+        }
+
+        yield return fileAnalysisData;
+    }
+
+    private IFileAnalysisData? Analyze(GameConstants targetGame, FileAnalyzerInfo info)
+    {
+        try
+        {
+            using var readStream = new MutagenInterfaceReadStream(
+                new BinaryReadStream(info.Stream, dispose: false),
+                new ParsingBundle(targetGame, masterReferences: null!)
+                {
+                    ModKey = ModKey.Null
+                }
+            );
+
+            var frame = new MutagenFrame(readStream);
+            var header = readStream.ReadModHeaderFrame(readSafe: true);
+
+            var isLightMaster = (header.Flags & (int)SkyrimModHeader.HeaderFlag.LightMaster) ==
+                                (int)SkyrimModHeader.HeaderFlag.LightMaster;
+
+            var masters = header
+                .Masters()
+                .Select(masterPinFrame =>
+                {
+                    readStream.Position = masterPinFrame.Location;
+                    return MasterReference.CreateFromBinary(frame);
+                })
+                .Select(masterReference => masterReference.Master.FileName.ToString().ToRelativePath())
+                .ToArray();
+
+            return new PluginAnalysisData
+            {
+                IsLightMaster = isLightMaster,
+                Masters = masters
+            };
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(e, "Exception while parsing {} ({})", info.FileName, info.RelativePath);
+            return null;
+        }
     }
 }

